src/btree/bt_split.c

```diff
@@ -1509,30 +1517,47 @@ __split_multi_inmem(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *multi, WT
  *     Discard moved update lists from the original page.
  */
 static void
-__split_multi_inmem_final(WT_PAGE *orig, WT_MULTI *multi)
+__split_multi_inmem_final(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *multi)
 {
     WT_SAVE_UPD *supd;
+    WT_UPDATE *upd;
     uint32_t i, slot;
 
+    /* If we have saved updates, we must have decided to restore them to the new page. */
+    WT_ASSERT(session, multi->supd_entries == 0 || multi->supd_restore);
+
     /*
      * We successfully created new in-memory pages. For error-handling reasons, we've left the
      * update chains referenced by both the original and new pages. We're ready to discard the
      * original page, terminate the original page's reference to any update list we moved.
      */
-    for (i = 0, supd = multi->supd; i < multi->supd_entries; ++i, ++supd)
-        switch (orig->type) {
-        case WT_PAGE_COL_FIX:
-        case WT_PAGE_COL_VAR:
+    for (i = 0, supd = multi->supd; i < multi->supd_entries; ++i, ++supd) {
+        /* We have finished restoration. Discard the update chains that aren't restored. */
+        if (!supd->restore)
+            continue;
```

Comment:
Where is the memory associated with this saved update that isn't being restored freed?

Reply:
It is freed along with the old page.

Reply:
Yes, you can do that by checking if onpage_upd is the first non-aborted transaction. Here we have the flag already so no need to walk the clean update chains.

Reply:
If we do that, we need to exclude in-memory and fixed length column store here as well.